onnxruntime/core/session/inference_session.cc

```diff
@@ -1733,59 +1740,111 @@ common::Status InferenceSession::Run(IOBinding& io_binding) {
+  std::future<void> setup_future = setup_promise.get_future();
+
+  // Passing run_options and io_binding by reference to the bg_thread,
+  // this is ok because they are ORTModule's member, and they are presistent through forward and backward calls
+  auto bg_thread = std::thread([this](std::future<void> setup_future, const RunOptions& run_options, IOBinding& io_binding) {
+    // wait until task is properly setup
+    setup_future.get();
+
+    common::Status status = Run(run_options, io_binding.GetInputNames(), io_binding.GetInputs(), io_binding.GetOutputNames(),
+                                &io_binding.GetOutputs(), &io_binding.GetOutputsDeviceInfo());
+
+    onnxruntime::contrib::OrtTasks::GetInstance().SetStatus(status);
+
+    // If forward outputs still hasn't been consumed at this point, i.e. forward function hasn't complete itself
+    // this indicates that Run() call returned before hitting YieldOp, due to hitting some exception during the forward subgraph execution
+    // In this case, we need to wake up the foreground thread and pass along the failed status.
+    // Otherwise, foreground thread will be stuck waiting for forward_outputs.
+    if (onnxruntime::contrib::OrtTasks::GetInstance().ForwardOutputsIsValid()) {
+      ORT_ENFORCE(!status.IsOK());
+      // signal main thread for background thread completion
+      onnxruntime::contrib::OrtTasks::GetInstance().SetForwardOutputs(status, {});
+    }
   },
-                                 std::move(task_.bg_thread_promise_));
+                               std::move(setup_future), std::cref(run_options), std::ref(io_binding));
 
-  // Wait for events from
+  run_id = std::hash<std::thread::id>()(bg_thread.get_id());
+  {
+    std::lock_guard<std::mutex> lock(bg_threads_mutex_);
+    bg_threads_[run_id] = std::move(bg_thread);
+  }
+
+  onnxruntime::contrib::OrtTasks::GetInstance().CreateBackgroundTask(run_id);
+
+  LOGS(*session_logger_, VERBOSE) << "InferenceSession::Forward() call created a task with run_id " << run_id;
+
+  // background task is setup, unblock background thread to continue
+  setup_promise.set_value();
+
+  // Wait for data/signal from
   // 1. Yield op, if the bg thread sucessfully reached Yield's signal point
   // 2. The end of bg thread, if it hit execptions and returned earlier
-  onnxruntime::contrib::OrtEventPool::GetInstance().WaitAndResetEvent(main_thread_event_id);
+  auto forward_outputs = onnxruntime::contrib::OrtTasks::GetInstance().WaitForForwardOutputs(run_id);
+  const Status& forward_status = forward_outputs.first;
+  user_outputs = std::move(forward_outputs.second);
 
   // background thread has completed without hitting Yield Op
-  if (task_.bg_thread_future_.wait_for(std::chrono::milliseconds(0)) == std::future_status::ready) {
-    Status bg_thread_status = task_.bg_thread_future_.get();
-    task_.bg_thread_.join();
-    return bg_thread_status;
+  if (!forward_status.IsOK()) {
+    std::thread bg_thread;
```

Comment:
I think there's a legit build breakage here because `bg_thread` shadows another local variable: https://dev.azure.com/onnxruntime/onnxruntime/_build/results?buildId=319324&view=logs&j=7b8df2e9-0609-506d-58f6-0285fd799216&t=a8eea653-37c6-59ae-bdc4-f0aafe8e474c&l=1539

I'd actually suggest moving this little snippet to a private method (e.g. `RemoveAndJoinBackgroundThread(run_id)`) since I think there are three places where the same logic appears in this file.